test/unit/org/apache/cassandra/db/DirectoriesTest.java

```diff
@@ -696,4 +732,144 @@ private List<Directories.DataDirectoryCandidate> getWriteableDirectories(DataDir
         return candidates;
     }
 
+    private int getDiyThreadId()
+    {
+        return myDiyId = diyThreadId.getAndIncrement();
+    }
+
+    private void detachLogger()
+    {
+        logger.detachAppender(listAppender);
+        MDC.remove(this.MDCID);
+    }
+
+    private void tailLogs()
+    {
+        int diyId = getDiyThreadId();
+        MDC.put(this.MDCID, String.valueOf(diyId));
+        logger = (Logger) LoggerFactory.getLogger(Directories.class);
+
+        // create and start a ListAppender
+        listAppender = new ListAppender<>();
+        listAppender.start();
+
+        // add the appender to the logger
+        logger.addAppender(listAppender);
+    }
+
+    private List<ILoggingEvent> filterLogByDiyId(List<ILoggingEvent> log)
+    {
+        ArrayList<ILoggingEvent> filteredLog = new ArrayList<>();
+        for(ILoggingEvent event : log)
+        {
+            int mdcId = Integer.parseInt(event.getMDCPropertyMap().get(this.MDCID));
+            if(mdcId == myDiyId){
+                filteredLog.add(event);
+            }
+        }
+        return filteredLog;
+    }
+
+    private void checkFormattedMessage(List<ILoggingEvent> log, Level expectedLevel, String expectedMessage, int expectedCount)
+    {
+        int found=0;
+        for(ILoggingEvent e: log)
+        {
+            System.err.println(e.getFormattedMessage());
+            if(e.getFormattedMessage().endsWith(expectedMessage))
+            {
+                if (e.getLevel() == expectedLevel)
+                    found++;
+            }
+        }
+
+        assertEquals(expectedCount, found);
+    }
+
+    @Test
+    public void testHasAvailableDiskSpace()
+    {
+        DataDirectory[] dataDirectories = new DataDirectory[]
```

Comment:
we may make a Parameterized test for three paramters: nearlyFullDir1, uniformDir2, veryFullDir
,This will make the logic clearer and the code more concise。
As line 817 to 873 have sane logic for these three paramters

Reply:
Yes. Thanks for saying this, I was hoping to discuss alternatives.

Your proposal will even make it easier to write and maintain these tests. The reason I didn't do that already is that it's unclear to me how this will affect the triaging of CI results. If you have a unit test like:

    for ( value : hundredValues)
        assertEquals(value, funtionToTest());

Now, if you get an assertion, how will you know which of the 100 values is the wrong one? JUnite will just show you the line number of that line, but the failure could be any of the 100 values we are looping over.

If you could suggest  a way to do that and also maintain great usability for those who run and ultimately have to fix it() anyway.

Reply:
I saw the ut test case in trunk and Parameterized test is created for sstable id for SequenceBasedSSTableId and UUIDBasedSSTableId. 

I just modify the code :
`    public static final DataDirectory uniformDir = new DataDirectory(new File("/uniformDir2"))
    {
        public long getAvailableSpace()
        {
            return 999L;
        }
    };
    public static final DataDirectory nearlyFullDir = new DataDirectory(new File("/nearlyFullDir1"))
    {
        public long getAvailableSpace()
        {
            return 11L;
        }
    };
    public static final DataDirectory veryFullDir = new DataDirectory(new File("/veryFullDir"))
    {
        public long getAvailableSpace()
        {
            return 4L;
        }
    };

 @Parameterized.Parameter
    public DataDirectory directory;

    @Parameterized.Parameter(1)
    publicLong value;

    @Parameterized.Parameter(2)
    public String message;

    @Parameterized.Parameters
    public static Collection<Object[]> idBuilders()
    {
        return Arrays.asList(new Object[]{ uniformDir, 999L, "uniformDir2" },
                                          new Object[]{ nearlyFullDir, 11L, "nearlyFullDir1" },
                                          new Object[]{ veryFullDir, 4L,"veryFullDir" });
    }
`

so code logic only needs to be there once, and duplicate logic can be deleted, then every time we can just verify the realation between directory, value, message.

Reply:
Hmm... I don't think that's equivalent to the test as submitted in the PR. First of all the dataDirectories should all exist together in each test. They aren't different input values for a parameterized test.

From there, note that the lines 

        String logMsgFormat = "DataDirectory %s has %d bytes available, checking if we can write %d bytes";
        String logMsg = String.format(logMsgFormat, "/nearlyFullDir1", 11, 2);
        checkFormattedMessage(filteredLog, Level.DEBUG, logMsg, 1);

Are merely checking log output from a test that already ran. If we wanted to parameterize, we can, but that would be just these 4 lines

        assertTrue(d.hasAvailableDiskSpace(1,2));
        assertTrue(d.hasAvailableDiskSpace(10,99));
        assertFalse(d.hasAvailableDiskSpace(10,1024));
        assertFalse(d.hasAvailableDiskSpace(1024,1024*1024));

And in the log output checking code, only 1 value could be parameterized. (The number 2 in what I quoted above.)

In short,I don't see the benefit of Parameterized here. The question remains: If I wrap the assertions into a simple loop:

        Collection<Object[]> checkTheseOut = Arrays.asList(new Object[]{ Level.DEBUG, "/nearlyFullDir1", 11, 2},
                             new Object[]{ "/uniformDir2, 999, 2 },
                              ...);
        for ( values: checkTheseOut )
        {
              String logMsg = String.format(logMsgFormat, (String)values[1], (int)values[2], (int)values[3]);
              checkFormattedMessage(filteredLog, (Level)values[0], logMsg, 1);
        }


...then how will the future developer running this unit test easily interpret the results when something failed?